clicommand/tool_sign.go

```diff
@@ -88,70 +130,201 @@ editor in the Buildkite UI so that the agents running these steps can verify the
+	key jwk.Key,
+	cfg *ToolSignConfig,
+) error {
+	if cfg.Repository == "" {
+		return ErrUseGraphQL
+	}
 
-			// Actually read the file from STDIN
-			input = os.Stdin
-			filename = "(stdin)"
+	// Find the pipeline either from STDIN or the first argument
+	var (
+		input    io.Reader
+		filename string
+	)
 
-		default:
-			return ErrNoPipeline
-		}
+	switch {
+	case cfg.FilePath != "":
+		l.Info("Reading pipeline config from %q", cfg.FilePath)
 
-		// Parse the pipeline
-		result, err := pipeline.Parse(input)
+		file, err := os.Open(cfg.FilePath)
 		if err != nil {
-			return fmt.Errorf("pipeline parsing of %q failed: %v", filename, err)
+			return fmt.Errorf("failed to read file: %w", err)
 		}
+		defer file.Close()
 
-		l.Debug("Pipeline parsed successfully: %v", result)
+		input = file
+		filename = cfg.FilePath
 
-		pInv := &pipeline.PipelineInvariants{
-			OrganizationSlug: cfg.OrganizationSlug,
-			PipelineSlug:     cfg.PipelineSlug,
-			Repository:       cfg.Repository,
-		}
+	case stdin.IsReadable():
+		l.Info("Reading pipeline config from STDIN")
 
-		key, err := loadSigningKey(&cfg)
-		if err != nil {
-			return fmt.Errorf("couldn't read the signing key file: %w", err)
+		input = os.Stdin
+		filename = "(stdin)"
+
+	default:
+		return ErrNoPipeline
+	}
+
+	parsedPipeline, err := pipeline.Parse(input)
+	if err != nil {
+		return fmt.Errorf("pipeline parsing of %q failed: %v", filename, err)
+	}
+
+	if cfg.Debug {
+		enc := yaml.NewEncoder(c.App.Writer)
+		enc.SetIndent(yamlIndent)
+		if err := enc.Encode(parsedPipeline); err != nil {
+			return fmt.Errorf("couldn't encode pipeline: %w", err)
 		}
+		l.Debug("Pipeline parsed successfully:\n%v", parsedPipeline)
+	}
+
+	if err := parsedPipeline.Sign(key, cfg.Repository); err != nil {
+		return fmt.Errorf("couldn't sign pipeline: %w", err)
+	}
+
+	enc := yaml.NewEncoder(c.App.Writer)
+	enc.SetIndent(yamlIndent)
+	return enc.Encode(parsedPipeline)
+}
 
-		if err := result.Sign(key, pInv); err != nil {
-			return fmt.Errorf("couldn't sign pipeline: %w", err)
+func signWithGraphQL(
+	ctx context.Context,
+	c *cli.Context,
+	l logger.Logger,
+	key jwk.Key,
+	cfg *ToolSignConfig,
+) error {
+	orgPipelineSlug := fmt.Sprintf("%s/%s", cfg.OrganizationSlug, cfg.PipelineSlug)
+	debugL := l.WithFields(logger.StringField("orgPipelineSlug", orgPipelineSlug))
+
+	l.Info("Retrieving pipeline from the GraphQL API")
+
+	client := bkgql.NewClient(cfg.GraphQLToken)
+
+	resp, err := bkgql.GetPipeline(ctx, client, orgPipelineSlug)
+	if err != nil {
+		return fmt.Errorf("couldn't retrieve pipeline: %w", err)
+	}
+
+	if resp.Pipeline.Id == "" {
+		return fmt.Errorf(
+			"%w: organization-slug: %s, pipeline-slug: %s",
+			ErrNotFound,
+			cfg.OrganizationSlug,
+			cfg.PipelineSlug,
+		)
+	}
+
+	debugL.Debug("Pipeline retrieved successfully: %#v", resp)
+	l.Info("Signing pipeline with the repository URL:\n%s", resp.Pipeline.Repository.Url)
+
+	pipelineYaml := strings.NewReader(resp.Pipeline.Steps.Yaml)
```

Comment:
YAML is an initialism, so I'd write it as `pipelineYAML`, even if the generated code has other ideas. Repeat below (`signedPipelineYAML`, `signedPipelineYAMLBuilder`)

```suggestion
	pipelineYAML := strings.NewReader(resp.Pipeline.Steps.Yaml)
```

Reply:
I've side stepped this by eliminating the variable. I've never liked names like `pipelineYAML`. It's bordering on Hungarian notation.

Reply:
Whoops, I did that in the other PR: https://github.com/buildkite/agent/pull/2481